File: src/main/java/org/trp/shincolle/network/S2CAdmiralDataSyncPayload.java
package org.trp.shincolle.network;

import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import org.trp.shincolle.Shincolle;

import java.util.HashSet;

public record S2CAdmiralDataSyncPayload(CompoundTag admiralNbt, int[] collectedShips) implements CustomPacketPayload {
    public static final Type<S2CAdmiralDataSyncPayload> TYPE = new Type<>(ResourceLocation.fromNamespaceAndPath(Shincolle.MODID, "s2c_admiral_data_sync"));

    public static final StreamCodec<FriendlyByteBuf, S2CAdmiralDataSyncPayload> STREAM_CODEC = StreamCodec.composite(
            ByteBufCodecs.COMPOUND_TAG, S2CAdmiralDataSyncPayload::admiralNbt,
            ByteBufCodecs.VAR_INT.apply(ByteBufCodecs.list()).map(list -> list.stream().mapToInt(Integer::intValue).toArray(), ints -> java.util.Arrays.stream(ints).boxed().toList()), S2CAdmiralDataSyncPayload::collectedShips,
            S2CAdmiralDataSyncPayload::new
    );

    public static S2CAdmiralDataSyncPayload of(CompoundTag admiralNbt, HashSet<Integer> collectedShips) {
        int[] values = collectedShips.stream().mapToInt(Integer::intValue).toArray();
        return new S2CAdmiralDataSyncPayload(admiralNbt, values);
    }

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return TYPE;
    }
}

File: src/main/java/org/trp/shincolle/network/S2CAdmiralDataSyncPayload.kt
package org.trp.shincolle.network

import io.netty.buffer.ByteBuf
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.ResourceLocation
import org.trp.shincolle.Shincolle
import java.util.*
import java.util.function.BiFunction
import java.util.function.Function

@JvmRecord
data class S2CAdmiralDataSyncPayload(val admiralNbt: CompoundTag?, val collectedShips: IntArray?) :
    CustomPacketPayload {
    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload?> {
        return TYPE
    }

    companion object {
        val TYPE: CustomPacketPayload.Type<S2CAdmiralDataSyncPayload?> =
            CustomPacketPayload.Type<S2CAdmiralDataSyncPayload?>(
                ResourceLocation.fromNamespaceAndPath(
                    Shincolle.MODID,
                    "s2c_admiral_data_sync"
                )
            )

        val STREAM_CODEC: StreamCodec<FriendlyByteBuf?, S2CAdmiralDataSyncPayload?> =
            StreamCodec.composite<FriendlyByteBuf?, S2CAdmiralDataSyncPayload?, CompoundTag?, IntArray?>(
                ByteBufCodecs.COMPOUND_TAG,
                S2CAdmiralDataSyncPayload::admiralNbt,
                ByteBufCodecs.VAR_INT.apply<MutableList<Int?>?>(ByteBufCodecs.list<ByteBuf?, Int?>())
                    .map<IntArray?>(Function { list: MutableList<Int?>? ->
                        list!!.stream().mapToInt { obj: Int? -> obj!!.toInt() }.toArray()
                    }, Function { ints: IntArray? -> Arrays.stream(ints).boxed().toList() }),
                S2CAdmiralDataSyncPayload::collectedShips,
                BiFunction { admiralNbt: CompoundTag?, collectedShips: IntArray? ->
                    S2CAdmiralDataSyncPayload(
                        admiralNbt,
                        collectedShips
                    )
                }
            )

        fun of(admiralNbt: CompoundTag?, collectedShips: HashSet<Int?>): S2CAdmiralDataSyncPayload {
            val values = collectedShips.stream().mapToInt { obj: Int? -> obj!!.toInt() }.toArray()
            return S2CAdmiralDataSyncPayload(admiralNbt, values)
        }
    }
}
